### PR TITLE
Review: Alternative implementation of handling forking

### PR DESCRIPTION
**Pull request by heifner**
I was concerned about the performance of `block_has_trx`.

Review comments for #966

### CODE REVIEW

File: plugins/trace_api_plugin/store_provider.cpp [hunk 1]
```diff
@@ -160,24 +159,21 @@ namespace eosio::trace_api {
...
                   }
                }
+               // block can be seen again when a fork happens, if not in the new block remove it from blocks that have the trx
+               if (!found_in_block)
+                  trx_block_nums.erase(trxs_entry.block_num);
             } else if (std::holds_alternative<lib_entry_v0>(entry)) {
                auto lib = std::get<lib_entry_v0>(entry).lib;
-               if (trx_entries > 0 && lib >= trx_block_num) {
-                  return trx_block_num;
+               if (!trx_block_nums.empty() && lib >= *(--trx_block_nums.end())) {
```
Comment (linh2931) [inline plugins/trace_api_plugin/store_provider.cpp:175]: Add a comment saying *(--trx_block_nums.end()) is the highest block number as trx_block_nums is a set of uint32_t.

File: plugins/trace_api_plugin/test/test_trace_file.cpp [hunk 2]
```diff
@@ -1146,21 +1151,22 @@ BOOST_AUTO_TEST_SUITE(slice_tests)
...
       // on_accepted_block of the initial block
-      sp.append(initial_block_trace);
-      sp.append_trx_ids(initial_block_trxs_entry);
+      sp.append(initial_block_trace);                  // block 1
+      sp.append_trx_ids(initial_block_trxs_entry);     // block 1
 
       // initial block forks out
-      sp.append(forked_block_trace);
+      sp.append(forked_block_trace);                   // block 1
+      sp.append_trx_ids(forked_block_trxs_entry);      // block 1
```
Comment (linh2931) [inline plugins/trace_api_plugin/test/test_trace_file.cpp:1159]: Thanks for this. I missed it.

**Review by linh2931 — APPROVED** [review]
Thanks! It does not need an extra look up the block's transaction list. Please go ahead to merge it into my branch. I will remove `block_has_trx()` and add additional tests.

**Comment by ericpassmore** [comment]
Note:start
category: System Stability
component: Internal
summary: Alternative implementation of handling forking.
Note:end

**Pull request merged**